### PR TITLE
Fix `replace_sampler` missing the batch size under specific conditions

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Reported in Slack.

Fixes a bug where the `batch_size` was not properly reset under the following conditions:
- `batch_size` is provided to the `DataLoader`
- No `batch_sampler` is present initially
- The `DataLoader` signature passes the `batch_size` through `**kwargs`.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by justusschock — APPROVED** [review]
LGTM. Can you add a simple test with a dummy loader?

Something like

```
class DummyLoader(torch.utils.data.DataLoader):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
```

**Review by SeanNaren — APPROVED** [review]

**Comment by SeanNaren** [comment]
> LGTM. Can you add a simple test with a dummy loader?
> 
> Something like
> 
> ```
> class DummyLoader(torch.utils.data.DataLoader):
>     def __init__(self, **kwargs):
>         super().__init__(**kwargs)
> ```

May be mistaken, but the `CustomDataLoader` test already does this:

```
class CustomDataLoader(DataLoader):
        def __init__(self, num_features, dataset, *args, **kwargs):
            self.num_features = num_features
            super().__init__(dataset, *args, **kwargs)
```

**Comment by justusschock** [comment]
@SeanNaren my bad, overseen this

File: pytorch_lightning/trainer/data_loading.py [hunk 1]
```diff
@@ -201,6 +201,11 @@ def _get_dataloader_init_kwargs(
...
         # get the dataloader instance `__init__` parameters
         params = dict(inspect.signature(dataloader.__init__).parameters)
+        has_variadic_kwargs = any(p.kind is p.VAR_KEYWORD for p in params.values())
+        if has_variadic_kwargs:
+            # if the signature takes **kwargs, assume they will be passed down with `super().__init__(**kwargs)`
+            params.update(inspect.signature(DataLoader.__init__).parameters)
+            del params["self"]
 
         # keep only the params whose default is different to the current attr value
         non_defaults = {name for name, p in params.items() if name in attrs and p.default != attrs[name]}
```
Comment (tchaton) [inline pytorch_lightning/trainer/data_loading.py:211]: Small note: This would break if default is an array.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/data_loading.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/data_loading.py:211]: Mutable objects are not supposed to be used as signature defaults
